### PR TITLE
Feature/mc 9125 Doc comments and types for key models

### PR DESCRIPTION
**Pull request by pjmonks**
A large PR with refactoring and additions to `mdm-resources` to provide better help with documentation and understanding of the endpoints. Because the number of endpoints is so large, this PR only focuses on the following key areas so far (with some additions):

* Data models
* Terminologies
* Code sets
* Reference data models

The following overall changes have been made:

1. Type definitions have been added to this library to provide the basework for client applications. For instance, `mdm-ui` will be modified to use these provided types to define data models instead of defining its own
2. Existing types have been renamed to provide consistency with TypeScript/JavaScript style e.g. `IMdmQueryStringParams` becomes `QueryParameters`
3. Doc comments have been added to a number of endpoint resources, with some cross-references. Doc comments then get used in IntelliSense in the downstream client application, plus are part of the `TypeDoc` generated API documentation.

There is an equivalent branch in `mdm-ui` called [feature/mc-9125](https://github.com/MauroDataMapper/mdm-ui/tree/feature/mc-9125) which has also been refactored to test these changes allow the UI to compile and work as expected.

**Note**: after this PR has been reviewed but before the PR is merged, the `mdm-ui` branch will also need review to avoid breaking changes to the UI repo.

### CODE REVIEW

File: src/mdm-common.model.ts [hunk 1]
```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2020 University of Oxford
```
Comment (gammonpeter) [inline src/mdm-common.model.ts:2]: 2021 copyright

File: src/mdm-data-class.model.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2020 University of Oxford
```
Comment (gammonpeter) [inline src/mdm-data-class.model.ts:2]: 2021

File: src/mdm-data-element.model.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+/*
+Copyright 2020 University of Oxford
```
Comment (gammonpeter) [inline src/mdm-data-element.model.ts:2]: 2021

File: src/mdm-provider.model.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
+/*
+Copyright 2020 University of Oxford
```
Comment (gammonpeter) [inline src/mdm-provider.model.ts:2]: 2021

File: src/mdm-reference-data-model.model.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
+/*
+Copyright 2020 University of Oxford
```
Comment (gammonpeter) [inline src/mdm-reference-data-model.model.ts:2]: 2021

File: src/mdm-session.model.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2020 University of Oxford
```
Comment (gammonpeter) [inline src/mdm-session.model.ts:2]: 2021

File: src/mdm-term.resource.ts [hunk 7]
```diff
@@ -39,79 +40,133 @@ import { IMdmQueryStringParams, IMdmRestHandlerOptions } from './mdm-rest-handle
...
+    return this.simpleGet(url, queryStringParams, restHandlerOptions);
+  }
+
+  /**
+   * `HTTP POST` - Creates a new term under a chosen terminology.
+   *
+   * @param terminologyId The unique identifier of the terminology to create the term under.
+   * @param data The payload of the request containing all the details for the term to create.
+   * @param query Optional query parameters to control the creation of the term, if required.
+   * @param options Optional REST handler parameters, if required.
```
Comment (gammonpeter) [inline src/mdm-term.resource.ts:67]: mentions query and options on the params list there are none in the actual method

**Review by gammonpeter — CHANGES_REQUESTED** [review]
Amazing work, just some small issues with copyright and mismatch of the description and params

File: src/mdm-admin.model.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
+/*
+Copyright 2020 University of Oxford
```
[the remark this answers is not in the document]
Comment (pjmonks) [reply src/mdm-admin.model.ts:2]: Fixed

File: src/mdm-common.model.ts (hunk 1 above)
Comment (pjmonks) [reply src/mdm-common.model.ts:2]: Fixed

File: src/mdm-data-element.model.ts (hunk 3 above)
Comment (pjmonks) [reply src/mdm-data-element.model.ts:2]: Fixed

File: src/mdm-provider.model.ts (hunk 4 above)
Comment (pjmonks) [reply src/mdm-provider.model.ts:2]: Fixed

File: src/mdm-reference-data-model.model.ts (hunk 5 above)
Comment (pjmonks) [reply src/mdm-reference-data-model.model.ts:2]: Fixed

File: src/mdm-term.model.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
+/*
+Copyright 2020 University of Oxford
```
[the remark this answers is not in the document]
Comment (pjmonks) [reply src/mdm-term.model.ts:2]: Fixed

File: src/mdm-term.resource.ts (hunk 7 above)
Comment (pjmonks) [reply src/mdm-term.resource.ts:67]: Fixed

File: src/mdm-terminology.model.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
+/*
+Copyright 2020 University of Oxford
```
[the remark this answers is not in the document]
Comment (pjmonks) [reply src/mdm-terminology.model.ts:2]: Fixed

File: src/mdm-tree-item.model.ts [hunk 11]
```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2020 University of Oxford
```
[the remark this answers is not in the document]
Comment (pjmonks) [reply src/mdm-tree-item.model.ts:2]: Fixed

**Pull request merged**